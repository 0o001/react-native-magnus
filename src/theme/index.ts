export { ThemeContext } from './themeContext';
export { withTheme } from './withTheme';
export { ThemeProvider, ThemeProviderProps } from './themeProvider.component';

export {
  ThemeType,
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from './type';
