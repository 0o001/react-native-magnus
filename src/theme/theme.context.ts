import * as React from 'react';

import { ThemeType } from './type';

const defaultValue: ThemeType = {};

export interface ThemeContextType {
  theme: ThemeType;
  setTheme?: (theme?: ThemeType) => void;
}

export const THEME_CONTEXT_DEFAULT_VALUE = {
  theme: defaultValue,
};

export const ThemeContext: React.Context<ThemeContextType> = React.createContext(
  THEME_CONTEXT_DEFAULT_VALUE
);
