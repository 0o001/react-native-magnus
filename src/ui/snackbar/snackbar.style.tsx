import { StyleSheet } from "react-native";

import {
  getThemeProperty,
  createSpacingStyles,
  createBorderColorStyles,
  createBorderWidthStyles,
  createBorderRadiusStyles
} from "../../theme/theme.service";

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.wrapper = {
    position: "absolute",
    bottom: 0,
    width: "100%"
  };

  computedStyle.text = {
    flex: 1,
    fontSize: getThemeProperty(theme.fontSize, props.fontSize, 16),
    color: getThemeProperty(theme.colors, props.color, "white")
  };

  computedStyle.prefix = {
    marginRight: 8
  };

  computedStyle.suffix = {
    marginLeft: 8
  };

  computedStyle.container = {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme.spacing),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius)
  };

  if (props.shadow) {
    computedStyle.container = {
      ...computedStyle.container,
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, "white")
    };
  }

  if (props.bg) {
    computedStyle.container = {
      ...computedStyle.container,
      backgroundColor: getThemeProperty(theme.colors, props.bg, "transparent")
    };
  }

  computedStyle.button = {
    marginHorizontal: 8,
    marginVertical: 6
  };

  return StyleSheet.create(computedStyle);
};
