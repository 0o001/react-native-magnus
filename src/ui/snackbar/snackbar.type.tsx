import { StyleProp, ViewStyle } from 'react-native';

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
  ShadowPropsType,
} from '../../theme';

export interface SnackbarRef {
  show: () => void;
  hide: () => void;
}

export interface SnackbarProps
  extends BorderPropsType,
    SpacingPropsType,
    ShadowPropsType,
    RoundedPropsType {
  bg?: string;
  color?: string;
  duration?: number;
  fontSize?: string;
  onDismiss?: () => void;
  children: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  suffix?: React.ReactNode;
  prefix?: React.ReactNode;
  bottom?: number;
  left?: number;
  top?: number;
  right?: number;
  w?: number;
  h?: number;
  minH?: number | string;
  minW?: number | string;
  maxH?: number | string;
  maxW?: number | string;
  opacity?: number;
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline';
  flexDir?: 'row' | 'column' | 'row-reverse' | 'column-reverse';
  alignSelf?:
    | 'auto'
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'stretch'
    | 'baseline';
  position?: 'absolute' | 'relative';
  useNativeDriver?: boolean;
}
