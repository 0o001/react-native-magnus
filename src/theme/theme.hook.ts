import React, { useState } from "react";

import { ThemeType } from "./type";
import { computeTheme } from "./theme.service";
import { ThemeContextType } from "./theme.context";

export const useTheme = (initialTheme?: ThemeType): ThemeContextType => {
  const [theme, setTheme] = useState(computeTheme(initialTheme));

  const setCurrentTheme = React.useCallback((newTheme?: ThemeType): void => {
    setTheme(computeTheme(newTheme));
  }, []);

  return {
    theme,
    setTheme: setCurrentTheme
  };
};
