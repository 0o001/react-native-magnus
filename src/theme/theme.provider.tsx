import * as React from "react";

import { ThemeType } from "./type";

import { ThemeContext } from "./theme.context";
import { useTheme } from "./theme.hook";

export interface ThemeProviderProps {
  theme?: ThemeType;
  children?: React.ReactNode;
}

export const ThemeProvider: React.FunctionComponent<ThemeProviderProps> = (
  props: ThemeProviderProps
) => {
  const { theme: themeProp, children } = props;
  const theme = useTheme(themeProp);

  return (
    <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
  );
};
