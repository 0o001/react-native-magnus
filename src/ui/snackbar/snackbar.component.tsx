import * as React from 'react';
import { useContext, useState, useEffect, useImperativeHandle } from 'react';
import { Animated, SafeAreaView, View as RNView } from 'react-native';

import { getStyle } from './snackbar.style';
import { ThemeContext } from '../../theme';
import { Text } from '../text/text.component';
import { SnackbarRefType, SnackbarProps } from './snackbar.type';

const DURATION_MEDIUM = 7000;

const Snackbar = React.forwardRef<SnackbarRefType, SnackbarProps>(
  (props, ref) => {
    const {
      m,
      mt,
      mr,
      mb,
      ml,
      p,
      pr,
      pt,
      pb,
      pl,
      rounded,
      roundedTop,
      roundedRight,
      roundedBottom,
      roundedLeft,
      children,
      onDismiss,
      duration,
      suffix,
      prefix,
      borderColor,
      borderBottomColor,
      borderLeftColor,
      borderTopColor,
      borderRightColor,
      borderWidth,
      borderLeftWidth,
      borderRightWidth,
      borderBottomWidth,
      borderTopWidth,
      ...rest
    } = props;
    const theme = useContext(ThemeContext);
    const computedStyle = getStyle(theme, props);
    const [opacity] = useState(new Animated.Value(0.0));
    const [hidden, setHidden] = useState(true);
    let hideTimeout: number;

    /**
     * component lifecycle methods
     */
    useEffect(() => {
      return function cleanup() {
        clearTimeout(hideTimeout);
      };
    }, []);

    /**
     * hide the snackbar
     */
    const hide = () => {
      if (hideTimeout) {
        clearTimeout(hideTimeout);
      }

      Animated.timing(opacity, {
        toValue: 0,
        duration: 100,
        useNativeDriver: true,
      }).start(({ finished }) => {
        if (finished) {
          setHidden(true);
        }
      });
    };

    /**
     * show snackbar
     */
    const show = () => {
      if (hideTimeout) {
        clearTimeout(hideTimeout);
      }

      setHidden(false);

      Animated.timing(opacity, {
        toValue: 1,
        duration: 200,
        useNativeDriver: true,
      }).start(({ finished }) => {
        if (finished) {
          const isInfinity =
            duration === Number.POSITIVE_INFINITY ||
            duration === Number.NEGATIVE_INFINITY;

          if (finished && !isInfinity) {
            hideTimeout = setTimeout(() => {
              hide();

              if (onDismiss) {
                onDismiss();
              }
            }, duration);
          }
        }
      });
    };

    /**
     * exposing functions to parent
     */
    useImperativeHandle(ref, () => ({
      show() {
        show();
      },
      hide() {
        hide();
      },
    }));

    // if snackbakr is set to be hidden, just return null
    if (hidden) {
      return null;
    }

    return (
      <SafeAreaView pointerEvents="box-none" style={computedStyle.wrapper}>
        <Animated.View
          style={{
            ...computedStyle.container,
            opacity: opacity,
            transform: [
              {
                scale: !hidden
                  ? opacity.interpolate({
                      inputRange: [0, 1],
                      outputRange: [0.9, 1],
                    })
                  : 1,
              },
            ],
          }}
          {...rest}>
          {prefix && <RNView style={computedStyle.prefix}>{prefix}</RNView>}
          <Text style={computedStyle.text}>{children}</Text>
          {suffix && <RNView style={computedStyle.suffix}>{suffix}</RNView>}
        </Animated.View>
      </SafeAreaView>
    );
  },
);

Snackbar.defaultProps = {
  bg: 'white',
  color: 'gray900',
  p: 'md',
  m: 'md',
  rounded: 'md',
  fontSize: 'text400',
  duration: DURATION_MEDIUM,
  onDismiss: () => {},
  shadow: 2,
  shadowColor: 'gray500',
};

export { Snackbar };
