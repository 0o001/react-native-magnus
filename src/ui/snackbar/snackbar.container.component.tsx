import React, { Component } from 'react';
import {
  StyleSheet,
  ViewStyle,
  KeyboardAvoidingView,
  Platform,
} from 'react-native';

import { Snackbar } from './snackbar.component';
import {
  SnackbarContainerState,
  SnackbarContainerProps,
  SnackbarProps,
} from './snackbar.type';

class SnackbarContainer extends Component<
  SnackbarContainerProps,
  SnackbarContainerState
> {
  constructor(props: SnackbarContainerProps) {
    super(props);

    this.state = {
      snackbars: [],
    };
  }

  static defaultProps = {
    placement: 'bottom',
  };

  show = (message: string | JSX.Element, config?: SnackbarProps) => {
    let id = config?.id || Math.random().toString();
    const onClose = () => this.hide(id);

    requestAnimationFrame(() => {
      this.setState({
        snackbars: this.state.snackbars.filter((t) => t.id !== id),
      });
      this.setState({
        snackbars: [
          {
            id,
            message,
            onClose,
            ...config,
          },
          ...this.state.snackbars,
        ],
      });
    });

    return id;
  };

  update = (
    id: string,
    message: string | JSX.Element,
    config?: SnackbarProps
  ) => {
    this.setState({
      snackbars: this.state.snackbars.map((toast) =>
        toast.id === id ? { ...toast, message, ...config } : toast
      ),
    });
  };

  hide = (id: string) => {
    this.setState({
      snackbars: this.state.snackbars.filter((t) => t.id !== id),
    });
  };

  render() {
    const { snackbars } = this.state;
    let { placement } = this.props;

    let style: ViewStyle = {
      justifyContent: placement === 'bottom' ? 'flex-end' : 'flex-start',
      flexDirection: placement === 'bottom' ? 'column' : 'column-reverse',
    };

    return (
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'position' : undefined}
        style={[styles.container, style]}
        pointerEvents="box-none"
      >
        {snackbars.map((snackbar) => {
          const { message, id, ...rest } = snackbar;

          return (
            <Snackbar key={id} {...this.props} {...rest}>
              {message}
            </Snackbar>
          );
        })}
      </KeyboardAvoidingView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 0,
    position: 'absolute',
    bottom: 0,
    left: 0,
    width: '100%',
  },
  message: {
    color: '#333',
  },
});

export { SnackbarContainer };
