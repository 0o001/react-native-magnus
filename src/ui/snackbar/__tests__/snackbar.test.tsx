import React from "react";
import renderer from "react-test-renderer";

import { Snackbar } from "../snackbar.component";
import { ThemeProvider } from "../../../theme";

describe("Snackbar Component", () => {
  it("should render without issues", () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Snackbar>Snack</Snackbar>
      </ThemeProvider>
    );

    expect(instance.toJSON()).toMatchSnapshot();

    const typeInst = instance.root.findByType(Snackbar);
    expect(typeInst.props.children).toBe("Snack");
  });
});
