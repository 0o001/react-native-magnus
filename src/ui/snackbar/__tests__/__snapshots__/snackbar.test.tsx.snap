// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar Component should render without issues 1`] = `null`;
